File: handlers/pjsk/b30.go
package pjsk

import (
	"encoding/json"
	"fmt"
	htmltemplate "html/template"
	"io"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/xiaocaoooo/amiabot-pages/pkg/imgcache"
)

var (
	pjskB30HTTPClient = &http.Client{Timeout: 30 * time.Second}
	b30ChartCache     = struct {
		mu      sync.RWMutex
		data    string
		expires time.Time
	}{}
)

// suiteMusicResult 来自 suite-api userMusicResults 的单条记录
type suiteMusicResult struct {
	MusicID        int    `json:"musicId"`
	MusicDifficulty string `json:"musicDifficultyType"`
	HighScore      int64  `json:"highScore"`
	PlayResult     string `json:"playResult"`
	FullComboFlg   bool   `json:"fullComboFlg"`
	FullPerfectFlg bool   `json:"fullPerfectFlg"`
	PlayType       string `json:"playType"`
}

// B30ChartEntry 歌曲难度 Constant 数据
type B30ChartEntry struct {
	SongID   int
	Diff     string
	Level    float64
	Constant float64
}

// B30ScoreEntry 单个成绩（模板渲染）
type B30ScoreEntry struct {
	Order      int
	Cover      htmltemplate.URL
	Name       string
	SongID     string
	Diff       string // MA/APD/...
	Level      string
	ResultType string // ap / fc / clear
	ResultIcon htmltemplate.URL
	Constant   string
	DiffStyle  htmltemplate.CSS
}

// B30PageData 传递给模板的数据
type B30PageData struct {
	Name           string
	UserID         string
	Server         string
	ServerKey      string
	UserRating     string
	Count          int
	Scores         []B30ScoreEntry
	ChartUpdatedAt string
	UpdatedTime    string
}

const b30ChartURL = "https://raw.githubusercontent.com/moe-sekai/MoeSekai-Hub/main/data/pjskb30/merged_chart.csv"

// ===== helpers =====

func diffNameToNum(diff string) int {
	switch strings.ToLower(diff) {
	case "easy":
		return 0
	case "normal":
		return 1
	case "hard":
		return 2
	case "expert":
		return 3
	case "master":
		return 4
	case "append":
		return 5
	default:
		return -1
	}
}

func diffNumToLabel(n int) string {
	switch n {
	case 0:
		return "EZ"
	case 1:
		return "NM"
	case 2:
		return "HD"
	case 3:
		return "EX"
	case 4:
		return "MA"
	case 5:
		return "APD"
	default:
		return "?"
	}
}

func diffNumToBGStyleHTML(n int) htmltemplate.CSS {
	switch n {
	case 0:
		return "background:#5AC06E;color:#fff;"
	case 1:
		return "background:#56A4D4;color:#fff;"
	case 2:
		return "background:#EFAF28;color:#fff;"
	case 3:
		return "background:#E84D53;color:#fff;"
	case 4:
		return "background:#BB58B8;color:#fff;"
	case 5:
		return "background:#EE92BC;color:#fff;"
	default:
		return "background:#888;color:#fff;"
	}
}

func formatB30(v float64) string { return fmt.Sprintf("%.2f", v) }

func truncateStr(s string, maxLen int) string {
	runes := []rune(s)
	if len(runes) > maxLen {
		return string(runes[:maxLen])
	}
	return s
}

func formatPJSKTimestamp(ts int64) string {
	if ts <= 0 {
		return ""
	}
	if ts < 100000000000 {
		return time.Unix(ts, 0).Local().Format("2006-01-02 15:04:05")
	}
	return time.UnixMilli(ts).Local().Format("2006-01-02 15:04:05")
}

func shortenErrBody(body []byte) string {
	text := strings.TrimSpace(string(body))
	if text == "" {
		return "空响应"
	}
	text = strings.ReplaceAll(text, "\n", " ")
	text = strings.ReplaceAll(text, "\r", " ")
	text = strings.Join(strings.Fields(text), " ")
	if len(text) > 240 {
		return text[:240] + "..."
	}
	return text
}

// ===== chart =====

func getB30ChartCSV() (string, error) {
	b30ChartCache.mu.RLock()
	if time.Now().Before(b30ChartCache.expires) && b30ChartCache.data != "" {
		data := b30ChartCache.data
		b30ChartCache.mu.RUnlock()
		return data, nil
	}
	b30ChartCache.mu.RUnlock()

	b30ChartCache.mu.Lock()
	defer b30ChartCache.mu.Unlock()

	if time.Now().Before(b30ChartCache.expires) && b30ChartCache.data != "" {
		return b30ChartCache.data, nil
	}

	resp, err := pjskB30HTTPClient.Get(b30ChartURL)
	if err != nil {
		if b30ChartCache.data != "" {
			return b30ChartCache.data, nil
		}
		return "", fmt.Errorf("获取难度表失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		if b30ChartCache.data != "" {
			return b30ChartCache.data, nil
		}
		return "", fmt.Errorf("读取难度表失败: %w", err)
	}

	b30ChartCache.data = string(body)
	b30ChartCache.expires = time.Now().Add(1 * time.Hour)
	return b30ChartCache.data, nil
}

func parseB30Chart(csv string) map[int]*B30ChartEntry {
	result := make(map[int]*B30ChartEntry)
	lines := strings.Split(csv, "\n")
	for i, line := range lines {
		if i == 0 || strings.TrimSpace(line) == "" {
			continue
		}
		fields := strings.SplitN(line, ",", 8)
		if len(fields) < 7 {
			continue
		}

		constant, err := strconv.ParseFloat(strings.TrimSpace(fields[2]), 64)
		if err != nil || constant <= 0 {
			continue
		}
		songID, err := strconv.Atoi(strings.TrimSpace(fields[6]))
		if err != nil || songID <= 0 {
			continue
		}

		diff := strings.TrimSpace(fields[5])
		diffNum := diffNameToNum(diff)
		if diffNum < 0 {
			continue
		}

		levelStr := strings.TrimSpace(fields[3])
		level := parseLevel(levelStr)
		if level <= 0 {
			continue
		}

		key := songID*10 + diffNum
		if existing, ok := result[key]; !ok || constant > existing.Constant {
			result[key] = &B30ChartEntry{SongID: songID, Diff: diff, Level: level, Constant: constant}
		}
	}
	return result
}

func parseLevel(s string) float64 {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0
	}
	s = strings.TrimRight(s, "+-")
	parts := strings.Fields(s)
	if len(parts) > 1 {
		s = parts[len(parts)-1]
	}
	level, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0
	}
	return level
}

// ===== rating calculation (lunabot style) =====

func resultTypeRank(playResult string, fullCombo, fullPerfect bool) int {
	if fullPerfect {
		return 3 // AP
	}
	if fullCombo {
		return 2 // FC
	}
	if playResult != "not_clear" {
		return 1 // Clear
	}
	return 0 // Not Clear
}

func calcRatingForResult(resultType int, constant float64, level float64) float64 {
	switch resultType {
	case 3: // AP
		return constant
	case 2: // FC
		if level >= 33 {
			return constant - 1.0
		}
		return constant - 1.5
	default:
		return 0
	}
}

// ===== fetch data =====

func loadSuiteBaseURL() string {
	return strings.TrimRight(strings.TrimSpace(os.Getenv("PJSK_SUITE_BASEURL")), "/")
}

func fetchSuiteMusicResults(suiteBaseURL, server, userID string) ([]suiteMusicResult, string, string, error) {
	endpoint := suiteBaseURL + "/public/" + server + "/suite/" + userID
	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, "", "", fmt.Errorf("创建 suite-api 请求失败: %w", err)
	}
	req.Header.Set("User-Agent", "amiabot-pages/1.0")
	req.Header.Set("Accept", "application/json")

	resp, err := pjskB30HTTPClient.Do(req)
	if err != nil {
		return nil, "", "", fmt.Errorf("请求 suite-api 失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, "", "", fmt.Errorf("suite-api 返回 %d: %s", resp.StatusCode, shortenErrBody(body))
	}

	var suiteData struct {
		UserMusicResults []suiteMusicResult `json:"userMusicResults"`
		UploadTime       int64              `json:"upload_time"`
		UserProfile      struct {
			UserId json.Number `json:"userId"`
			Name   string      `json:"name"`
		} `json:"userProfile"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&suiteData); err != nil {
		return nil, "", "", fmt.Errorf("解析 suite-api 响应失败: %w", err)
	}

	if len(suiteData.UserMusicResults) == 0 {
		return nil, "", "", fmt.Errorf("suite-api 未返回 userMusicResults 数据")
	}

	var name string
	if suiteData.UserProfile.Name != "" {
		name = strings.TrimSpace(suiteData.UserProfile.Name)
	} else {
		if profileCfg, err := loadPJSKProfileAPIConfigFromEnv(); err == nil {
			name = fetchB30PlayerName(profileCfg, server, userID)
		}
	}
	if name == "" {
		userIDStr := suiteData.UserProfile.UserId.String()
		if userIDStr != "" {
			name = "玩家 " + userIDStr
		} else {
			name = "玩家 " + userID
		}
	}

	var uploadTime string
	if suiteData.UploadTime > 0 {
		uploadTime = formatPJSKTimestamp(suiteData.UploadTime)
	}

	return suiteData.UserMusicResults, name, uploadTime, nil
}

func fetchB30PlayerName(cfg *pjskProfileAPIConfig, server, userID string) string {
	if cfg == nil || cfg.BaseURL == "" {
		return ""
	}
	endpoint := cfg.BaseURL + "/" + server + "/" + userID + "/profile"
	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
	if err != nil {
		return ""
	}
	req.Header.Set("User-Agent", "amiabot-pages/1.0")
	req.Header.Set("Accept", "application/json")
	for k, v := range cfg.Headers {
		req.Header.Set(k, v)
	}
	resp, err := pjskB30HTTPClient.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return ""
	}
	var result struct {
		User struct {
			Name string `json:"name"`
		} `json:"user"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ""
	}
	return strings.TrimSpace(result.User.Name)
}

// ===== music index (read musics.json once) =====

func buildMusicIndex(server string) map[int]*pjskMusic {
	data, err := ReadCachedJSON(server, "musics.json")
	if err != nil {
		return nil
	}
	var musics []pjskMusic
	if err := json.Unmarshal(data, &musics); err != nil {
		return nil
	}
	idx := make(map[int]*pjskMusic, len(musics))
	for i := range musics {
		idx[musics[i].ID] = &musics[i]
	}
	return idx
}

// ===== handler =====

func B30Handler(c *gin.Context) {
	server := c.DefaultQuery("server", "jp")
	if !validServers[server] {
		renderB30Err(c, "无效的服务器参数，支持: jp, cn, en, tw, kr")
		return
	}

	userID := strings.TrimSpace(c.Query("id"))
	if userID == "" {
		renderB30Err(c, "缺少玩家 ID 参数")
		return
	}
	if !isDigits(userID) {
		renderB30Err(c, "无效的玩家 ID: "+userID)
		return
	}

	suiteBaseURL := loadSuiteBaseURL()
	if suiteBaseURL == "" {
		renderB30Err(c, "未配置 PJSK_SUITE_BASEURL 环境变量")
		return
	}

	// 1. 获取难度表
	chartCSV, err := getB30ChartCSV()
	if err != nil {
		renderB30Err(c, "加载难度表失败: "+err.Error())
		return
	}
	chartMap := parseB30Chart(chartCSV)

	// 2. 获取音乐成绩
	musicResults, name, uploadTime, err := fetchSuiteMusicResults(suiteBaseURL, server, userID)
	if err != nil {
		renderB30Err(c, err.Error())
		return
	}

	// 3. 按 (musicId, difficulty) 取最佳成绩 (AP > FC > Clear > NotClear)
	type bestEntry struct {
		songID     int
		diffNum    int
		diffLabel  string
		level      float64
		constant   float64
		resultType int
		resultStr  string
	}

	bestMap := make(map[int]bestEntry)

	for _, r := range musicResults {
		songID := r.MusicID
		diffNum := diffNameToNum(strings.ToLower(r.MusicDifficulty))
		if songID <= 0 || diffNum < 0 {
			continue
		}

		chart, ok := chartMap[songID*10+diffNum]
		if !ok {
			continue
		}

		rt := resultTypeRank(r.PlayResult, r.FullComboFlg, r.FullPerfectFlg)
		if rt == 0 {
			continue
		}

		key := songID*10 + diffNum
		if existing, exists := bestMap[key]; !exists || rt > existing.resultType {
			var resultStr string
			switch rt {
			case 3:
				resultStr = "ap"
			case 2:
				resultStr = "fc"
			default:
				resultStr = "clear"
			}

			bestMap[key] = bestEntry{
				songID:     songID,
				diffNum:    diffNum,
				diffLabel:  diffNumToLabel(diffNum),
				level:      chart.Level,
				constant:   chart.Constant,
				resultType: rt,
				resultStr:  resultStr,
			}
		}
	}

	// 4. 计算每条 rating 并排序
	type scoredEntry struct {
		bestEntry
		rating float64
	}

	var scored []scoredEntry
	for _, entry := range bestMap {
		rating := calcRatingForResult(entry.resultType, entry.constant, entry.level)
		if rating <= 0 {
			continue
		}
		scored = append(scored, scoredEntry{bestEntry: entry, rating: rating})
	}

	sort.Slice(scored, func(i, j int) bool {
		return scored[i].rating > scored[j].rating
	})
	if len(scored) > 30 {
		scored = scored[:30]
	}

	sumRating := 0.0
	for _, s := range scored {
		sumRating += s.rating
	}

	serverName := serverNames[server]
	if serverName == "" {
		serverName = strings.ToUpper(server)
	}

	// 5. 一次性读取 musics.json，构建 title + cover 索引
	musicIdx := buildMusicIndex(server)

	// 6. 转换为模板视图
	views := make([]B30ScoreEntry, len(scored))
	for i, s := range scored {
		resultIcon := pjskB30ResultIcon(s.resultStr)

		constantText := strconv.FormatFloat(s.constant, 'f', -1, 64)

		var cover htmltemplate.URL
		var songName string
		if m := musicIdx[s.songID]; m != nil {
			songName = m.Title
			cover = downloadAssetByLabel(server, "music:jacket:"+m.AssetbundleName)
		}

		views[i] = B30ScoreEntry{
			Order:      i + 1,
			Cover:      cover,
			Name:       truncateStr(songName, 20),
			SongID:     "#" + strconv.Itoa(s.songID),
			Diff:       s.diffLabel,
			Level:      levelLabel(s.level),
			ResultType: s.resultStr,
			ResultIcon: resultIcon,
			Constant:   constantText,
			DiffStyle:  diffNumToBGStyleHTML(s.diffNum),
		}
	}

	page := B30PageData{
		Name:           name,
		UserID:         userID,
		Server:         serverName,
		ServerKey:      server,
		UserRating:     formatB30(sumRating / 30.0),
		Count:          len(views),
		Scores:         views,
		ChartUpdatedAt: b30ChartUpdateText(),
		UpdatedTime:    uploadTime,
	}

	c.HTML(http.StatusOK, "pjsk/b30", gin.H{"B30": page})
}

func renderB30Err(c *gin.Context, errMsg string) {
	c.HTML(http.StatusOK, "pjsk/b30", gin.H{"Error": errMsg})
}

func levelLabel(level float64) string {
	return strings.TrimRight(strings.TrimRight(fmt.Sprintf("%.0f", level), "0"), ".")
}

func b30ChartUpdateText() string {
	b30ChartCache.mu.RLock()
	defer b30ChartCache.mu.RUnlock()
	if b30ChartCache.data == "" {
		return ""
	}
	return ""
}

// ===== result icons =====

const (
	pjskB30APIconURL = "https://raw.githubusercontent.com/watagashi-uni/Unibot/refs/heads/main/pics/AllPerfect.png"
	pjskB30FCIconURL = "https://raw.githubusercontent.com/watagashi-uni/Unibot/refs/heads/main/pics/FullCombo.png"
)

func pjskB30ResultIcon(resultType string) htmltemplate.URL {
	switch resultType {
	case "ap":
		return imgcache.Default.Download(pjskB30APIconURL, -1, nil)
	case "fc":
		return imgcache.Default.Download(pjskB30FCIconURL, -1, nil)
	default:
		return ""
	}
}

File: src/handlers/pjsk/b30.rs
use axum::{extract::Query, response::IntoResponse};
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::RwLock;

use crate::handlers::pjsk::asset_source::download_asset_by_label;
use crate::handlers::pjsk::assets::read_cached_json;
use crate::handlers::pjsk::{SERVER_NAMES, VALID_SERVERS};
use crate::handlers::{format_upstream_http_error, render_html};
use crate::pkg::imgcache::DEFAULT_IMG_CACHE;

const B30_CHART_URL: &str =
    "https://raw.githubusercontent.com/moe-sekai/MoeSekai-Hub/main/data/pjskb30/merged_chart.csv";
const PJSK_B30_AP_ICON_URL: &str =
    "https://raw.githubusercontent.com/watagashi-uni/Unibot/refs/heads/main/pics/AllPerfect.png";
const PJSK_B30_FC_ICON_URL: &str =
    "https://raw.githubusercontent.com/watagashi-uni/Unibot/refs/heads/main/pics/FullCombo.png";

#[derive(Deserialize, Debug)]
pub struct B30Query {
    pub id: Option<String>,
    pub server: Option<String>,
}

#[derive(Serialize, Clone)]
#[allow(non_snake_case)]
pub struct B30ScoreEntry {
    pub Order: usize,
    pub Cover: String,
    pub Name: String,
    pub SongID: String,
    pub Diff: String,
    pub Level: String,
    pub ResultType: String,
    pub ResultIcon: String,
    pub Constant: String,
    pub DiffStyle: String,
}

#[derive(Serialize, Clone)]
#[allow(non_snake_case)]
pub struct B30PageData {
    pub Name: String,
    pub Server: String,
    pub ServerKey: String,
    pub UserRating: String,
    pub Count: usize,
    pub Scores: Vec<B30ScoreEntry>,
    pub ChartUpdatedAt: String,
    pub UpdatedTime: String,
    pub FooterExtra: String,
}

#[derive(Serialize)]
#[allow(non_snake_case)]
pub struct B30Response {
    pub B30: Option<B30PageData>,
    pub Error: Option<String>,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct MusicEntry {
    id: i32,
    title: String,
    assetbundleName: String,
}

struct ChartCache {
    data: String,
    expires: std::time::SystemTime,
}

static B30_CHART_CACHE: Lazy<Arc<RwLock<Option<ChartCache>>>> =
    Lazy::new(|| Arc::new(RwLock::new(None)));

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct SuiteUserMusicResult {
    musicId: i32,
    musicDifficultyType: String,
    highScore: i64,
    playResult: String,
    fullComboFlg: bool,
    fullPerfectFlg: bool,
    playType: String,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct SuiteUserProfile {
    name: Option<String>,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct SuiteUserMusicResponse {
    userMusicResults: Vec<SuiteUserMusicResult>,
    #[serde(default)]
    userProfile: Option<SuiteUserProfile>,
    #[serde(default)]
    upload_time: Option<i64>,
}

fn diff_name_to_num(diff: &str) -> i32 {
    match diff.to_lowercase().as_str() {
        "easy" => 0,
        "normal" => 1,
        "hard" => 2,
        "expert" => 3,
        "master" => 4,
        "append" => 5,
        _ => -1,
    }
}

fn diff_num_to_label(n: i32) -> &'static str {
    match n {
        0 => "EZ",
        1 => "NM",
        2 => "HD",
        3 => "EX",
        4 => "MA",
        5 => "APD",
        _ => "?",
    }
}

fn diff_num_to_bg_style(n: i32) -> &'static str {
    match n {
        0 => "background:#5AC06E;color:#fff;",
        1 => "background:#56A4D4;color:#fff;",
        2 => "background:#EFAF28;color:#fff;",
        3 => "background:#E84D53;color:#fff;",
        4 => "background:#BB58B8;color:#fff;",
        5 => "background:#EE92BC;color:#fff;",
        _ => "background:#888;color:#fff;",
    }
}

async fn get_b30_chart_csv() -> Result<String, String> {
    {
        let cache = B30_CHART_CACHE.read().await;
        if let Some(ref c) = *cache {
            if c.expires > std::time::SystemTime::now() {
                return Ok(c.data.clone());
            }
        }
    }

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .unwrap_or_default();

    let resp = crate::pkg::http_client::send(client.get(B30_CHART_URL))
        .await
        .map_err(|e| format!("获取难度表失败: {}", e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format_upstream_http_error("B30 难度表", status, &body));
    }

    let data = resp
        .text()
        .await
        .map_err(|e| format!("读取难度表失败: {}", e))?;

    let mut cache = B30_CHART_CACHE.write().await;
    *cache = Some(ChartCache {
        data: data.clone(),
        expires: std::time::SystemTime::now() + Duration::from_secs(3600),
    });

    Ok(data)
}

struct ChartEntry {
    level: f64,
    constant: f64,
}

fn parse_b30_chart(csv: &str) -> HashMap<i32, ChartEntry> {
    let mut result = HashMap::new();
    let lines = csv.split('\n');
    for (i, line) in lines.enumerate() {
        if i == 0 || line.trim().is_empty() {
            continue;
        }
        let fields: Vec<&str> = line.splitn(8, ',').collect();
        if fields.len() < 7 {
            continue;
        }

        let constant: f64 = match fields[2].trim().parse() {
            Ok(c) if c > 0.0 => c,
            _ => continue,
        };
        let song_id: i32 = match fields[6].trim().parse() {
            Ok(id) if id > 0 => id,
            _ => continue,
        };

        let diff = fields[5].trim();
        let diff_num = diff_name_to_num(diff);
        if diff_num < 0 {
            continue;
        }

        let level: f64 = fields[1].trim().parse().unwrap_or(0.0);

        result.insert(song_id * 10 + diff_num, ChartEntry { level, constant });
    }
    result
}

fn result_type_rank(play_result: &str, fc: bool, ap: bool) -> i32 {
    let lower = play_result.to_lowercase();
    if lower == "not_clear" {
        return 0;
    }
    if ap {
        return 3;
    }
    if fc {
        return 2;
    }
    1 // clear
}

fn calc_rating_for_result(rt: i32, constant: f64, _level: f64) -> f64 {
    if constant <= 0.0 {
        return 0.0;
    }
    match rt {
        3 => constant + 0.3, // AP
        2 => constant + 0.2, // FC
        1 => constant,       // Clear
        _ => 0.0,
    }
}

async fn fetch_suite_music_results(
    base_url: &str,
    server: &str,
    user_id: &str,
) -> Result<(Vec<SuiteUserMusicResult>, String, String), String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .unwrap_or_default();

    // Haruki suite-api: /public/{server}/suite/{user_id}
    let url = format!(
        "{}/public/{}/suite/{}",
        base_url.trim_end_matches('/'),
        server,
        user_id
    );

    let mut builder = client
        .get(&url)
        .header("User-Agent", "amiabot-pages/1.0")
        .header("Accept", "application/json");
    if let Ok(headers_str) = env::var("PJSK_PROFILE_HEADERS") {
        if let Ok(map) = serde_json::from_str::<HashMap<String, String>>(&headers_str) {
            for (k, v) in map {
                builder = builder.header(k, v);
            }
        }
    }

    let resp = crate::pkg::http_client::send(builder)
        .await
        .map_err(|e| format!("请求 suite-api 失败: {}", e))?;
    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format_upstream_http_error("B30 suite-api", status, &body));
    }

    let data = resp
        .json::<SuiteUserMusicResponse>()
        .await
        .map_err(|e| format!("解析 suite-api 成绩 JSON 失败: {}", e))?;

    if data.userMusicResults.is_empty() {
        return Err("suite-api 未返回 userMusicResults 数据".to_string());
    }

    let mut name = data
        .userProfile
        .and_then(|p| p.name)
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .unwrap_or_default();
    if name.is_empty() {
        name = crate::handlers::pjsk::profile::fetch_player_name(server, user_id)
            .await
            .unwrap_or_default();
    }
    if name.is_empty() {
        name = "玩家".to_string();
    }

    let upload_time = data
        .upload_time
        .map(|ts| {
            let secs = if ts < 100_000_000_000 { ts } else { ts / 1000 };
            crate::pkg::timefmt::format_unix_secs(secs)
        })
        .unwrap_or_default();

    Ok((data.userMusicResults, name, upload_time))
}

fn build_music_index(server: &str) -> HashMap<i32, MusicEntry> {
    let mut index = HashMap::new();
    if let Ok(data) = read_cached_json(server, "musics.json") {
        if let Ok(list) = serde_json::from_slice::<Vec<MusicEntry>>(&data) {
            for entry in list {
                index.insert(entry.id, entry);
            }
        }
    }
    index
}

fn render_b30_err(msg: &str) -> impl IntoResponse {
    tracing::warn!(error = %msg, "b30 页面错误");
    render_html(
        "pjsk/b30.html",
        B30Response {
            B30: None,
            Error: Some(msg.to_string()),
        },
    )
}

struct BestEntry {
    song_id: i32,
    diff_num: i32,
    diff_label: &'static str,
    level: f64,
    constant: f64,
    result_type: i32,
    result_str: &'static str,
}

pub async fn b30_handler(Query(q): Query<B30Query>) -> impl IntoResponse {
    let server = q.server.unwrap_or_else(|| "jp".to_string());
    if !VALID_SERVERS.contains(&server) {
        return render_b30_err("无效的服务器参数，支持: jp, cn, en, tw, kr").into_response();
    }

    let user_id = q.id.unwrap_or_default().trim().to_string();
    if user_id.is_empty() {
        return render_b30_err("缺少玩家 ID 参数").into_response();
    }
    if !user_id.chars().all(|c| c.is_ascii_digit()) {
        return render_b30_err("无效的玩家 ID").into_response();
    }

    let suite_base_url =
        match env::var("PJSK_SUITE_BASEURL").or_else(|_| env::var("PJSK_PROFILE_BASEURL")) {
            Ok(url) if !url.trim().is_empty() => url,
            _ => return render_b30_err("未配置 PJSK_SUITE_BASEURL 环境变量").into_response(),
        };

    // 1. 获取难度表
    let chart_csv = match get_b30_chart_csv().await {
        Ok(csv) => csv,
        Err(e) => return render_b30_err(&format!("加载难度表失败: {}", e)).into_response(),
    };
    let chart_map = parse_b30_chart(&chart_csv);

    // 2. 获取音乐成绩
    let (music_results, name, upload_time) =
        match fetch_suite_music_results(&suite_base_url, &server, &user_id).await {
            Ok(res) => res,
            Err(e) => return render_b30_err(&e).into_response(),
        };

    // 3. 并发合并最佳成绩
    let mut best_map: HashMap<i32, BestEntry> = HashMap::new();
    for r in music_results {
        let song_id = r.musicId;
        let diff_num = diff_name_to_num(&r.musicDifficultyType);
        if song_id <= 0 || diff_num < 0 {
            continue;
        }

        let chart = match chart_map.get(&(song_id * 10 + diff_num)) {
            Some(c) => c,
            None => continue,
        };

        let rt = result_type_rank(&r.playResult, r.fullComboFlg, r.fullPerfectFlg);
        if rt == 0 {
            continue;
        }

        let key = song_id * 10 + diff_num;
        let replace = match best_map.get(&key) {
            Some(existing) => rt > existing.result_type,
            None => true,
        };

        if replace {
            let result_str = match rt {
                3 => "ap",
                2 => "fc",
                _ => "clear",
            };
            best_map.insert(
                key,
                BestEntry {
                    song_id,
                    diff_num,
                    diff_label: diff_num_to_label(diff_num),
                    level: chart.level,
                    constant: chart.constant,
                    result_type: rt,
                    result_str,
                },
            );
        }
    }

    // 4. 计算 Rating 并排序
    struct ScoredEntry {
        entry: BestEntry,
        rating: f64,
    }

    let mut scored = Vec::new();
    for (_, entry) in best_map {
        let rating = calc_rating_for_result(entry.result_type, entry.constant, entry.level);
        if rating <= 0.0 {
            continue;
        }
        scored.push(ScoredEntry { entry, rating });
    }

    scored.sort_by(|a, b| {
        b.rating
            .partial_cmp(&a.rating)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    if scored.len() > 30 {
        scored.truncate(30);
    }

    let sum_rating: f64 = scored.iter().map(|s| s.rating).sum();
    let server_name = SERVER_NAMES
        .get(&server)
        .cloned()
        .unwrap_or_else(|| server.to_uppercase());
    let music_idx = build_music_index(&server);

    // 5. 转换为模板数据并异步拉取封面与结果图片
    let mut views = Vec::new();
    for (i, s) in scored.into_iter().enumerate() {
        let result_icon = match s.entry.result_str {
            "ap" => {
                DEFAULT_IMG_CACHE
                    .download(PJSK_B30_AP_ICON_URL, None, None)
                    .await
            }
            "fc" => {
                DEFAULT_IMG_CACHE
                    .download(PJSK_B30_FC_ICON_URL, None, None)
                    .await
            }
            _ => String::new(),
        };

        let constant_text = format!("{:.1}", s.entry.constant);
        let mut song_name = String::new();
        let mut cover = String::new();

        if let Some(m) = music_idx.get(&s.entry.song_id) {
            song_name = m.title.clone();
            let label = format!("music:jacket:{}", m.assetbundleName);
            cover = download_asset_by_label(&server, &label).await;
        }

        views.push(B30ScoreEntry {
            Order: i + 1,
            Cover: cover,
            Name: truncate_str(&song_name, 20),
            SongID: format!("#{}", s.entry.song_id),
            Diff: s.entry.diff_label.to_string(),
            Level: format!("{:.0}", s.entry.level),
            ResultType: s.entry.result_str.to_string(),
            ResultIcon: result_icon,
            Constant: constant_text,
            DiffStyle: diff_num_to_bg_style(s.entry.diff_num).to_string(),
        });
    }

    let rating_val = if views.is_empty() {
        0.0
    } else {
        sum_rating / 30.0
    };
    let page = B30PageData {
        Name: name,
        Server: server_name,
        ServerKey: server,
        UserRating: format!("{:.2}", rating_val),
        Count: views.len(),
        Scores: views,
        ChartUpdatedAt: String::new(), // Placeholder as Go version
        UpdatedTime: upload_time,
        FooterExtra: "Powered by Moesekai, Haruki, LunaBot, Uni, & Sekai World<br />".to_string(),
    };

    render_html(
        "pjsk/b30.html",
        B30Response {
            B30: Some(page),
            Error: None,
        },
    )
    .into_response()
}

fn truncate_str(s: &str, max_len: usize) -> String {
    let mut chars = s.chars();
    let mut res = String::new();
    for _ in 0..max_len {
        if let Some(c) = chars.next() {
            res.push(c);
        } else {
            break;
        }
    }
    if chars.next().is_some() {
        res.push('…');
    }
    res
}
